collect_app/src/androidTest/java/org/odk/collect/android/feature/smoke/BadServerTest.java

```diff
@@ -42,24 +41,16 @@ public OpenRosaHttpInterface provideHttpInterface(MimeTypeMap mimeTypeMap, UserA
             .around(rule);
 
     @Test
-    public void whenHashNotIncludedInFormList_canGetBlankForm_fillItIn_andSubmit() {
+    // The hash from the form list wasn't used for a long time so some server implementations omitted it even though
+    // it's required by the spec. Now we explicitly show an error.
+    public void whenHashNotIncludedInFormList_showError() {
         server.removeHashInFormList();
         server.addForm("One Question", "one-question", "1", "one-question.xml");
 
         rule.startAtMainMenu()
                 .setServer(server.getURL())
                 .clickGetBlankForm()
                 .clickGetSelected()
-                .assertText("One Question (Version:: 1 ID: one-question) - Success")
-                .clickOK(new MainMenuPage())
-
-                .startBlankForm("One Question")
-                .swipeToEndScreen()
-                .clickSaveAndExit()
-
-                .clickSendFinalizedForm(1)
-                .clickOnForm("One Question")
-                .clickSendSelected()
-                .assertText("One Question - Success");
+                .assertText("One Question (Version:: 1 ID: one-question) - Missing form hash. ODK-compatible servers must include form hashes in their form lists. Please talk to the person who asked you to collect data.");
```

Comment:
This should probably use the (translated) string resource for the assertion so that the test doesn't fail if there is a tweak to the language.

Reply:
https://github.com/getodk/collect/pull/4597/files#diff-bf920b275657bdd44fb5321255809b7ab723c2a2fe66b89706efd95ac5f3a7e9R82 -- it's not translated. I meant to bring this up in the PR description but now see it's only in the commit details. This is going to be short lived and it targets just a couple of rogue servers so I'd rather not have translators deal with it. This is especially important since it's really a message aimed at server developers. "form hash" probably shouldn't be translated, for example.

Reply:
For the legacy formlist request, I used the translated string for "Please talk..." mostly because that one will be more common. I can do this here if you prefer.

Reply:
Ah right sorry I missed that in the PR. Thinking about this a bit more, I feel a bit uncomfortable with the `DatabaseFormRepository` declaring the message. We'll need to readdress this whenever we move the `DatabaseFormsRepository` over to Kotlin as the method will no longer accept `null` at all after that, and that will push the message up. I think I'm happy to leave this in right now and avoid making changes to the use cases where we're more likely to see conflicts.

Reply:
Sounds good. We're going to have to rethink all of this sooner than later anyway because the generic `Failure` message in all other cases is causing user pain. I agree this is temporary.